### PR TITLE
Liquidity withdrawal

### PR DESCRIPTION
**Pull request by mantikoros**

### CODE REVIEW

File: common/calculate-cpmm.ts [hunk 1]
```diff
@@ -260,27 +260,34 @@ export function addCpmmLiquidity(
...
+
+  const liquidity = newLiquidity - oldLiquidity
+  return liquidity
+}
+
 export function getCpmmLiquidityPoolWeights(
   contract: CPMMContract,
-  liquidities: LiquidityProvision[]
+  liquidities: LiquidityProvision[],
+  excludeAntes = true
```
Comment (jahooma) [inline common/calculate-cpmm.ts:276]: Might be clearer to just remove this param, and always exclude antes, if that's how we use it. Makes the code simpler. Can change later to support including antes if necessary.

File: functions/src/withdraw-liquidity.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+      const result = await firestore.runTransaction(async (trans) => {
+        const lpDoc = firestore.doc(`users/${userId}`)
+        const lpSnap = await trans.get(lpDoc)
+        if (!lpSnap.exists) throw new APIError(400, 'User not found.')
+        const lp = lpSnap.data() as User
+
+        const contractDoc = firestore.doc(`contracts/${contractId}`)
+        const contractSnap = await trans.get(contractDoc)
+        if (!contractSnap.exists) throw new APIError(400, 'Contract not found.')
+        const contract = contractSnap.data() as CPMMContract
```
Comment (jahooma) [inline functions/src/withdraw-liquidity.ts:40]: Should you check if the contract is binary CPMM?

File: web/components/contract/contract-info-dialog.tsx [hunk 3]
```diff
@@ -113,14 +113,9 @@ export function ContractInfoDialog(props: { contract: Contract; bets: Bet[] }) {
...
 
-          {contract.mechanism === 'cpmm-1' &&
-            !contract.resolution &&
-            (!closeTime || closeTime > Date.now()) && (
-              <>
-                <div className="">Add liquidity</div>
-                <AddLiquidityPanel contract={contract} />
-              </>
-            )}
+          {contract.mechanism === 'cpmm-1' && !contract.resolution && (
```
Comment (jahooma) [inline web/components/contract/contract-info-dialog.tsx:116]: What about closed markets? You can still add / withdraw liquidity?

File: web/components/liquidity-panel.tsx [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+          content: <AddLiquidityPanel contract={contract} />,
+        },
+        ...(showWithdrawal
+          ? [
+              {
+                title: 'Withdraw liquidity',
+                content: (
+                  <WithdrawLiquidityPanel
+                    contract={contract}
+                    lpShares={lpShares as any}
```
Comment (jahooma) [inline web/components/liquidity-panel.tsx:42]: At least cast to { YES: number, NO: number }

**Review by jahooma — APPROVED** [review]
Looks pretty good! Cool new feature for our power users.

BTW, you should always include a screenshot in PRs that modify the UI.

File: web/components/contract/contract-info-dialog.tsx (hunk 3 above)
Comment (mantikoros) [reply web/components/contract/contract-info-dialog.tsx:116]: yes!

File: web/components/liquidity-panel.tsx (hunk 4 above)
Comment (mantikoros) [reply web/components/liquidity-panel.tsx:42]: ok

File: common/calculate-cpmm.ts (hunk 1 above)
Comment (mantikoros) [reply common/calculate-cpmm.ts:276]: ok

File: common/calculate-cpmm.ts [hunk 5]
```diff
@@ -260,27 +260,34 @@ export function addCpmmLiquidity(
   return { newPool, liquidity, newP }
 }
 
+const calculateContractLiquidity = (p: number) => (l: LiquidityProvision) => {
```
[the remark this answers is not in the document]
Comment (mantikoros) [reply common/calculate-cpmm.ts:263]: ok

**Pull request merged**